src/main/flight/pid.c

```diff
@@ -445,19 +446,62 @@ STATIC_UNIT_TESTED FAST_CODE_NOINLINE float pidLevel(int axis, const pidProfile_
 #ifdef USE_GPS_RESCUE
     angle += gpsRescueAngle[axis] / 100; // ANGLE IS IN CENTIDEGREES
 #endif
-    angle = constrainf(angle, -levelAngleLimit, levelAngleLimit);
     const float errorAngle = angle - ((attitude.raw[axis] - angleTrim->raw[axis]) / 10.0f);
+    const float levelPGainBoost = (1.0f + (fabsf(errorAngle) / levelAngleLimit));
+    float angleErrorCorrection = errorAngle * levelPGainBoost * pidRuntime.levelGain;
+    //angle loop feedforward
+    float angleFeedforward = 0;
+#ifdef USE_FEEDFORWARD
+    if (newRcFrame){
+        const float feedforwardJitterFactor = pidGetFeedforwardJitterFactor();
+        const float rcCommandDelta = fabsf(getRcCommandDelta(axis));
+        float angleFeedforwardInput = pidRuntime.angleFeedforward[axis];
+        uint8_t duplicateCount = getFeedforwardDuplicateCount(axis);
+        if (rcCommandDelta) {
+            const float rxRateHz = 1e6f / getCurrentRxRefreshRate();
+            angleFeedforwardInput = (angle - pidRuntime.previousAngle[axis]) * rxRateHz;
+            //jitter attenuation
+            float jitterAttenuator = 1.0f;
+            if (feedforwardJitterFactor) {
+                if (rcCommandDelta < feedforwardJitterFactor) {
+                    jitterAttenuator = MAX(1.0f - (rcCommandDelta / feedforwardJitterFactor), 0.0f);
+                    jitterAttenuator = 1.0f - jitterAttenuator * jitterAttenuator;
+                }
+            }
+            angleFeedforwardInput *= jitterAttenuator;
+            //duplicate handling
+            if (duplicateCount) {
+                    angleFeedforwardInput /= duplicateCount + 1;
+                }
+        } else {
+            if (duplicateCount > 1) {
+                angleFeedforwardInput = 0;
+            }
+        }
+        //angle feedforward filter at RCrate
+        pidRuntime.angleFeedforward[axis] = pt2FilterApply(&pidRuntime.angleFeedforwardPt2[axis], angleFeedforwardInput);
+        pidRuntime.previousAngle[axis] = angle;
+    }
+    //angle feedforward filter at the PID loop rate
+    angleFeedforward = pt3FilterApply(&pidRuntime.angleFeedforwardPt3[axis], pidRuntime.angleFeedforward[axis]);
+    angleFeedforward *= pidRuntime.levelFeedforwardGain;
```

Comment:
I think that probably in Horizon mode we won't want any angle feedforward.
That's because the pilot has gets acro setpoint control, and Horizon code adds some self-levelling.  The direct control is mostly supposed to come from acro mode.  So I think this multiplication can be done only in the Angle mode code.